File: resume-backend/app/repositories/applications.py
from __future__ import annotations

from datetime import date, datetime, timezone
import json
from pathlib import Path
from uuid import uuid4

from app.db import connect
from app.schemas.application import (
    ApplicationRecord,
    ApplicationSaveRequest,
    ApplicationStatus,
    InterviewReminderRequest,
    TimelineEvent,
    TimelineEventRequest,
)


class ApplicationNotFoundError(Exception):
    pass


class ApplicationRepository:
    def __init__(self, database_path: Path) -> None:
        self._database_path = database_path

    def list(
        self,
        user_id: str,
        status: ApplicationStatus | None = None,
        interview_date: date | None = None,
    ) -> list[ApplicationRecord]:
        query = """
            SELECT * FROM application_tracker
            WHERE user_id = ?
        """
        values: list[object] = [user_id]
        if status:
            query += " AND status = ?"
            values.append(status)
        if interview_date:
            query += " AND substr(next_interview_at, 1, 10) = ?"
            values.append(interview_date.isoformat())
        query += """
            ORDER BY
                CASE WHEN next_action_at IS NULL THEN 1 ELSE 0 END,
                next_action_at ASC,
                updated_at DESC,
                id DESC
        """
        with connect(self._database_path) as connection:
            rows = connection.execute(query, values).fetchall()
        return [self._from_row(row) for row in rows]

    def save(self, user_id: str, payload: ApplicationSaveRequest) -> ApplicationRecord:
        application_id = payload.id or str(uuid4())
        now = datetime.now(timezone.utc).isoformat()
        values = (
            payload.company,
            payload.role_name,
            payload.city,
            payload.source,
            payload.status,
            payload.applied_at.isoformat() if payload.applied_at else None,
            payload.next_action_at.isoformat() if payload.next_action_at else None,
            payload.interview_notes,
            payload.draft_id,
            payload.notes,
            payload.contact_info,
            payload.attachment_ref,
            payload.next_interview_at.isoformat() if payload.next_interview_at else None,
        )
        with connect(self._database_path) as connection:
            if payload.id:
                cursor = connection.execute(
                    """
                    UPDATE application_tracker
                    SET company = ?, role_name = ?, city = ?, source = ?, status = ?,
                        applied_at = ?, next_action_at = ?, interview_notes = ?,
                        draft_id = ?, notes = ?, contact_info = ?, attachment_ref = ?,
                        next_interview_at = ?, updated_at = ?
                    WHERE id = ? AND user_id = ?
                    """,
                    (*values, now, application_id, user_id),
                )
                if cursor.rowcount == 0:
                    raise ApplicationNotFoundError
            else:
                connection.execute(
                    """
                    INSERT INTO application_tracker (
                        id, client_id, user_id, company, role_name, city, source, status,
                        applied_at, next_action_at, interview_notes, draft_id, notes,
                        contact_info, attachment_ref, next_interview_at, created_at, updated_at
                    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                    """,
                    (application_id, user_id, user_id, *values, now, now),
                )
            row = connection.execute(
                "SELECT * FROM application_tracker WHERE id = ? AND user_id = ?",
                (application_id, user_id),
            ).fetchone()
        if row is None:
            raise ApplicationNotFoundError
        return self._from_row(row)

    def list_timeline(self, user_id: str, application_id: str) -> list[TimelineEvent]:
        with connect(self._database_path) as connection:
            row = self._owned_row(connection, user_id, application_id)
        return self._timeline_from_row(row)

    def add_timeline_event(
        self,
        user_id: str,
        application_id: str,
        payload: TimelineEventRequest,
    ) -> TimelineEvent:
        event = TimelineEvent(id=str(uuid4()), **payload.model_dump())
        with connect(self._database_path) as connection:
            row = self._owned_row(connection, user_id, application_id)
            timeline = self._timeline_from_row(row)
            timeline.append(event)
            connection.execute(
                """
                UPDATE application_tracker
                SET timeline_json = ?, updated_at = ?
                WHERE id = ? AND user_id = ?
                """,
                (
                    json.dumps([item.model_dump(mode="json") for item in timeline], ensure_ascii=False),
                    datetime.now(timezone.utc).isoformat(),
                    application_id,
                    user_id,
                ),
            )
        return event

    def save_reminder(
        self,
        user_id: str,
        application_id: str,
        payload: InterviewReminderRequest,
    ) -> dict[str, str]:
        reminder_id = str(uuid4())
        now = datetime.now(timezone.utc).isoformat()
        with connect(self._database_path) as connection:
            self._owned_row(connection, user_id, application_id)
            connection.execute(
                """
                INSERT INTO interview_reminder
                (id, user_id, application_id, reminder_at, status, created_at, updated_at)
                VALUES (?, ?, ?, ?, 'pending', ?, ?)
                """,
                (reminder_id, user_id, application_id, payload.reminder_at.isoformat(), now, now),
            )
        return {
            "id": reminder_id,
            "application_id": application_id,
            "reminder_at": payload.reminder_at.isoformat(),
            "status": "pending",
        }

    def delete(self, user_id: str, application_id: str) -> None:
        with connect(self._database_path) as connection:
            cursor = connection.execute(
                "DELETE FROM application_tracker WHERE id = ? AND user_id = ?",
                (application_id, user_id),
            )
        if cursor.rowcount == 0:
            raise ApplicationNotFoundError

    @staticmethod
    def _owned_row(connection, user_id: str, application_id: str):
        row = connection.execute(
            "SELECT * FROM application_tracker WHERE id = ? AND user_id = ?",
            (application_id, user_id),
        ).fetchone()
        if row is None:
            raise ApplicationNotFoundError
        return row

    @staticmethod
    def _timeline_from_row(row) -> list[TimelineEvent]:
        try:
            raw_timeline = json.loads(str(row["timeline_json"] or "[]"))
        except (TypeError, json.JSONDecodeError):
            return []
        if not isinstance(raw_timeline, list):
            return []
        events: list[TimelineEvent] = []
        for item in raw_timeline:
            if not isinstance(item, dict):
                continue
            try:
                events.append(TimelineEvent.model_validate(item))
            except ValueError:
                continue
        return events

    @staticmethod
    def _from_row(row) -> ApplicationRecord:
        return ApplicationRecord(
            id=str(row["id"]),
            company=str(row["company"]),
            role_name=str(row["role_name"]),
            city=str(row["city"]),
            source=str(row["source"]),
            status=str(row["status"]),
            applied_at=row["applied_at"],
            next_action_at=row["next_action_at"],
            interview_notes=str(row["interview_notes"]),
            draft_id=row["draft_id"],
            notes=str(row["notes"]),
            contact_info=str(row["contact_info"] or ""),
            attachment_ref=str(row["attachment_ref"] or ""),
            next_interview_at=row["next_interview_at"],
            timeline=ApplicationRepository._timeline_from_row(row),
            created_at=str(row["created_at"]),
            updated_at=str(row["updated_at"]),
        )

File: web-frontend/src/lib/applications.ts
import { requestApi } from "./api"

export type ApplicationStatus = "saved" | "applied" | "screening" | "interview" | "offer" | "rejected" | "closed"

export type ApplicationTimelineEvent = {
  id: string
  title: string
  description: string
  occurredAt: string
}

export type ApplicationRecord = {
  id: string
  company: string
  roleName: string
  city: string
  source: string
  status: ApplicationStatus
  appliedAt: string | null
  nextActionAt: string | null
  interviewNotes: string
  draftId: string | null
  notes: string
  contactInfo: string
  attachmentRef: string
  nextInterviewAt: string | null
  timeline: ApplicationTimelineEvent[]
  createdAt: string
  updatedAt: string
}

export type ApplicationInput = Omit<ApplicationRecord, "createdAt" | "updatedAt" | "timeline">

export type ApplicationFilters = {
  status?: ApplicationStatus
  interviewDate?: string
}

type BackendTimeline = {
  id: string
  title: string
  description: string
  occurred_at: string
}

type BackendApplication = {
  id: string
  company: string
  role_name: string
  city: string
  source: string
  status: ApplicationStatus
  applied_at: string | null
  next_action_at: string | null
  interview_notes: string
  draft_id: string | null
  notes: string
  contact_info?: string
  attachment_ref?: string
  next_interview_at?: string | null
  timeline?: BackendTimeline[]
  created_at: string
  updated_at: string
}

function fromTimeline(item: BackendTimeline): ApplicationTimelineEvent {
  return {
    id: item.id,
    title: item.title,
    description: item.description,
    occurredAt: item.occurred_at,
  }
}

function fromBackend(item: BackendApplication): ApplicationRecord {
  return {
    id: item.id,
    company: item.company,
    roleName: item.role_name,
    city: item.city,
    source: item.source,
    status: item.status,
    appliedAt: item.applied_at,
    nextActionAt: item.next_action_at,
    interviewNotes: item.interview_notes,
    draftId: item.draft_id,
    notes: item.notes,
    contactInfo: item.contact_info || "",
    attachmentRef: item.attachment_ref || "",
    nextInterviewAt: item.next_interview_at || null,
    timeline: (item.timeline || []).map(fromTimeline),
    createdAt: item.created_at,
    updatedAt: item.updated_at,
  }
}

function toBackend(input: ApplicationInput) {
  return {
    id: input.id,
    company: input.company,
    role_name: input.roleName,
    city: input.city,
    source: input.source,
    status: input.status,
    applied_at: input.appliedAt,
    next_action_at: input.nextActionAt,
    interview_notes: input.interviewNotes,
    draft_id: input.draftId,
    notes: input.notes,
    contact_info: input.contactInfo,
    attachment_ref: input.attachmentRef,
    next_interview_at: input.nextInterviewAt,
  }
}

export async function listApplications(filters: ApplicationFilters = {}): Promise<ApplicationRecord[]> {
  const params = new URLSearchParams()
  if (filters.status) params.set("status", filters.status)
  if (filters.interviewDate) params.set("interview_date", filters.interviewDate)
  const query = params.toString()
  const data = await requestApi<{ items: BackendApplication[] }>("/api/applications" + (query ? "?" + query : ""))
  return data.items.map(fromBackend)
}

export async function saveApplication(input: ApplicationInput): Promise<ApplicationRecord> {
  return fromBackend(await requestApi<BackendApplication>("/api/applications", {
    method: "POST",
    body: JSON.stringify(toBackend(input)),
  }))
}

export async function listTimeline(id: string): Promise<ApplicationTimelineEvent[]> {
  const data = await requestApi<{ items: BackendTimeline[] }>("/api/applications/" + encodeURIComponent(id) + "/timeline")
  return data.items.map(fromTimeline)
}

export async function addTimelineEvent(
  id: string,
  input: Omit<ApplicationTimelineEvent, "id">,
): Promise<ApplicationTimelineEvent> {
  return fromTimeline(await requestApi<BackendTimeline>("/api/applications/" + encodeURIComponent(id) + "/timeline", {
    method: "POST",
    body: JSON.stringify({
      title: input.title,
      description: input.description,
      occurred_at: input.occurredAt,
    }),
  }))
}

export async function saveReminder(id: string, reminderAt: string): Promise<void> {
  await requestApi("/api/applications/" + encodeURIComponent(id) + "/reminders", {
    method: "POST",
    body: JSON.stringify({ reminder_at: reminderAt }),
  })
}

export async function deleteApplication(id: string): Promise<void> {
  await requestApi<{ id: string }>("/api/applications/" + encodeURIComponent(id), { method: "DELETE" })
}
